### ISSUE TITLE
A `-single-line` or `-multi-line` option reads a block another rule is about to break over two lines

### ISSUE THREAD

**Issue by firefoxic** [stylelint-stylistic/stylelint-stylistic#355]
Half the options of this plugin ask whether the thing they are about is written on one line or on several — `always-single-line`, `never-single-line`, `always-multi-line`, `never-multi-line`. That question is asked of the text as it stands when the rule runs, and another rule in the same run may be about to write a line break into it. Sixty pairs of rules were measured to answer differently depending on which of the two the configuration lists first.

## The run

Fixture `@media screen{\na{b:c;d:e}\n}\n`, both rules at once, only the order changed:

```
block-opening-brace-space-before first          -> "@media screen {\na{b:c;\nd:e}\n}\n"    one warning left
declaration-block-semicolon-newline-after first -> "@media screen {\na {b:c;\nd:e}\n}\n"   nothing left to report
```

`block-opening-brace-space-before: "always-multi-line"` looks at the rule `a{b:c;d:e}`, sees one line, and says nothing. Then `declaration-block-semicolon-newline-after` breaks that block over two lines — and the space the first rule would now require is never written. It is not a text left unformatted: it is a rule that stayed silent about a block it would have spoken about a moment later.

The mirror case is just as easy to reach:

```
block-closing-brace-newline-before first  -> "@media screen{ a{ b:c;d:e\n}\n}\n"   both clean
block-opening-brace-space-after first     -> "@media screen{ a{b:c;d:e\n}\n}\n"    both clean
```

Here both outputs satisfy both rules, and they are still two different files. Two people with the same set of rules and a different order in their configuration format the same source differently.

## Why

`lintPostcssResult.mjs` of `stylelint@17.14.1` sorts the rules of a run by their place in Stylelint's own registry, and a plugin's rules are not in it:

```js
const rulesOrder = new Map(Object.keys(rules).map((ruleName, index) => [ruleName, index]))

ruleNames = Object.keys(config.rules).sort((a, b) => (rulesOrder.get(a) ?? -1) - (rulesOrder.get(b) ?? -1))
```

Every rule of this plugin resolves to `-1`, the sort is stable, and so the order is the one the user's configuration happens to spell. `lintPostcssResult` is called once per lint; there is no loop to a fixed point.

## The size of it

Sixty pairs, of which twenty-four reach the hard form — one order leaves a rule the user turned on unsatisfied, and reports nothing about it. The other thirty-six reach the soft form: both orders end up self-consistent, and the two files differ.

The rules taking part:

`block-closing-brace-empty-line-before`, `block-closing-brace-newline-before`, `block-closing-brace-space-before`, `block-opening-brace-newline-after`, `block-opening-brace-newline-before`, `block-opening-brace-space-after`, `block-opening-brace-space-before`, `declaration-block-semicolon-newline-after`, `declaration-block-semicolon-newline-before`, `declaration-block-semicolon-space-after`, `declaration-block-semicolon-space-before`, `declaration-colon-newline-after`, `declaration-colon-space-after`, `function-comma-newline-after`, `function-comma-newline-before`, `function-comma-space-after`, `function-comma-space-before`, `function-parentheses-newline-inside`, `function-parentheses-space-inside`, `selector-list-comma-newline-after`, `selector-list-comma-newline-before`, `selector-list-comma-space-after`, `selector-list-comma-space-before`.

## What ends it

This class does not yield to the remedy the three sibling classes take. There the fix is that a rule writing text writes it in its finished form; here nothing is written wrongly — a rule declines to speak, correctly, about a block that is about to stop being what it was.

Two directions are worth weighing before anything is written:

1. **Ask the question of the finished text.** A rule with a lineness option would have to know whether any other rule in the configuration is going to break this block. That is knowable — `result.stylelint.config` names every rule and its settings — but it means predicting another rule's fix rather than reading its setting.
2. **Ask the question of the source.** Lineness is measured once, off the text as the file spells it, and every rule in the run answers against that measurement. The order stops mattering, at the cost that a rule speaks about the block the file had rather than the block the run is producing.

The measurement ran 79 rules under 208 primary options over eight fixtures of plain CSS.

One of the four classes of [#356](https://github.com/stylelint-stylistic/stylelint-stylistic/issues/356), which is about the class rather than about any member of it.
